shuffle-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/ucx/UCXShuffleTransport.scala

```diff
@@ -148,6 +149,15 @@ class UCXShuffleTransport(shuffleServerId: BlockManagerId, rapidsConf: RapidsCon
         deviceNumBuffers,
         (size: Long) => DeviceMemoryBuffer.allocate(size))
 
+    // we need to hook onto the bounce buffer manager, since now frees are
+    // encapsulated in a `BounceBuffer`, but we need visibility to notify
+    // a monitor
+    deviceReceiveBuffMgr.onFree(() => {
+      inflightMonitor.synchronized {
+        inflightMonitor.notifyAll()
```

Comment:
Normally code that performs a wakeup also modifies something with the lock held.  Modifying the sleep conditional state without the lock held can lead to indefinite sleeps (a problem which is partially mitigated by using a sleep timeout as a workaround).